sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/GeneratedColumn.scala

```diff
@@ -0,0 +1,144 @@
+      fieldName: String,
+      schema: StructType,
+      statementType: String): String = {
+    // Parse the expression string
+    val parsed: Expression = try {
+      session.sessionState.sqlParser.parseExpression(expressionStr)
+    } catch {
+      case ex: ParseException =>
+        // Shouldn't be possible since we check that the expression is a valid catalyst expression
+        // during parsing
+        throw new AnalysisException(
+          s"Failed to execute $statementType command because the column $fieldName has " +
+            s"generation expression $expressionStr which fails to parse as a valid expression:" +
+            s"\n${ex.getMessage}")
+    }
+    // Analyze the parse result
+    // Generated column can't reference itself
+    val relation = new LocalRelation(StructType(schema.filterNot(_.name == fieldName)).toAttributes)
+    val plan = try {
+      val analyzer: Analyzer = GeneratedColumnAnalyzer
+      val analyzed = analyzer.execute(Project(Seq(Alias(parsed, fieldName)()), relation))
+      analyzer.checkAnalysis(analyzed)
+      analyzed
+    } catch {
+      case ex: AnalysisException =>
+        val columnList = schema.filterNot(_.name == fieldName).map(_.name).mkString("[", ",", "]")
+        throw new AnalysisException(
+          s"Failed to execute $statementType command because the column $fieldName has " +
+            s"generation expression $expressionStr which fails to resolve as a valid expression " +
+            s"given columns $columnList:" +
+            s"\n${ex.getMessage}")
+    }
+    val analyzed = plan.collectFirst {
+      case Project(Seq(a: Alias), _: LocalRelation) => a.child
+    }.get
+    // Try to convert to V2 Expression and then to SQL string
+    new V2ExpressionBuilder(analyzed).build().getOrElse {
+      throw new AnalysisException(
+        errorClass = "UNSUPPORTED_EXPRESSION_GENERATED_COLUMN",
+        messageParameters = Map(
+          "fieldName" -> fieldName,
+          "expressionStr" -> expressionStr
+        )
+      )
+    }.toString // toString uses V2ExpressionSQLBuilder
+  }
+
+  /**
+   * For any generated columns in `schema`, verify that the generation expression is a valid
+   * V2 [[org.apache.spark.sql.connector.expressions.Expression]] and convert the expression string
+   * to V2 Expression SQL.
+   */
+  def verifyAndConvertToV2ExpressionSQL(session: SparkSession,
+    schema: StructType, statementType: String): StructType = {
+    val newFields = schema.map { field =>
+      getGenerationExpression(field).map { expressionStr =>
+        val updatedExpressionStr =
+          convertToV2ExpressionSQL(session, expressionStr, field.name, schema, statementType)
+        field.copy(
+          metadata = new MetadataBuilder().withMetadata(field.metadata)
+            .putString(Table.GENERATION_EXPRESSION_METADATA_KEY, updatedExpressionStr)
+            .build()
+        )
+      }.getOrElse(field)
+    }
+    StructType(newFields)
+  }
+}
+
+/**
```

Comment:
This just uses the `BuiltInFunctionCatalog` defined in `ResolveDefaultColumnsUtil`. We could factor these out into a separate utils to just hold the `BuiltInFunctionCatalog` and a "`BuiltInFuctionAnalyzer`"